src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value_impl.hpp

```diff
@@ -0,0 +1,457 @@
+            typename boost::enable_if<IsVector<VecType>>*) const
+{
+  if (!HasValue())
+    return -1;
+
+  return CompareValues(localDataset->col(numValues - 1),*valueToInsert);
+}
+
+template<typename TreeElemType>
+template<typename TreeType, typename VecType>
+size_t DiscreteHilbertValue<TreeElemType>::
+InsertPoint(TreeType *node, const VecType& pt,
+                            typename boost::enable_if<IsVector<VecType>>*)
+{
+  size_t i = 0;
+
+  // All point are inserted to the root node
+  if (!node->Parent())
+    *valueToInsert = CalculateValue(pt);
+  if (node->IsLeaf())
+  {
+    // Find an appropriate place
+    for (i = 0; i < numValues; i++)
+      if (CompareValues(localDataset->col(i), *valueToInsert) > 0)
+        break;
+
+    for (size_t j = numValues; j > i; j--)
+      localDataset->col(j) = localDataset->col(j-1);
+
+    localDataset->col(i) = *valueToInsert;
+    numValues++;
+    // Propogate changes of the largest Hilbert value downward
+    TreeType* root = node->Parent();
+
+    while (root != NULL)
+    {
+      root->AuxiliaryInfo().HilbertValue().UpdateLargestValue(root);
+
+      root = root->Parent();
+    }
+
+  }
+
+  return i;
+}
+
+template<typename TreeElemType>
+template<typename TreeType>
+void DiscreteHilbertValue<TreeElemType>::InsertNode(TreeType* node)
+{
+  DiscreteHilbertValue &val = node->AuxiliaryInfo().HilbertValue();
+  
+  if (CompareWith(node,val) < 0)
+  {
+    localDataset = val.LocalDataset();
+    numValues = val.NumValues();
+  }
+}
+
+template<typename TreeElemType>
+template<typename TreeType>
+void DiscreteHilbertValue<TreeElemType>::
+DeletePoint(TreeType* node, const size_t localIndex)
+{
+
+  // Delete the Hilbert value from the local dataset
+  for (size_t i = numValues - 1; i > localIndex; i--)
+    localDataset->col(i-1) = localDataset->col(i);
+
+  numValues--;
+}
+
+template<typename TreeElemType>
+template<typename TreeType>
+void DiscreteHilbertValue<TreeElemType>::
+RemoveNode(TreeType* node, const size_t nodeIndex)
+{
+  if (node->NumChildren() <= 1)
+  {
+    localDataset = NULL;
+    numValues = 0;
+    return;
+  }
+  if (nodeIndex + 1 == node->NumChildren())
+  {
+    // Update the largest Hilbert value if the value exists
+    TreeType* child = node->Children()[nodeIndex-1];
+    if (child->AuxiliaryInfo.HilbertValue().NumValues() != 0)
+    {
+      numValues = child->AuxiliaryInfo.HilbertValue().NumValues();
+      localDataset = child->AuxiliaryInfo.HilbertValue().LocalDataset();
+    }
+    else
+    {
+      localDataset = NULL;
+      numValues = 0;
+    }
+  }
+}
+
+template<typename TreeElemType>
+template<typename TreeType>
+void DiscreteHilbertValue<TreeElemType>::Copy(TreeType* dst, TreeType* src)
+{
+  DiscreteHilbertValue<TreeElemType> &dstVal = dst->AuxiliaryInfo().HilbertValue();
+  DiscreteHilbertValue<TreeElemType> &srcVal = src->AuxiliaryInfo().HilbertValue();
+
+  dst.LocalDataset() = src.LocalDataset();
+  dst.NumValues() = src.NumValues();
```

Comment:
I think that there can be weird memory issues when calling this method because `ownsLocalDataset` is not considered.  I think it may be better to remove this method and replace it with a copy constructor or a move constructor (or both); the same comments apply to `HilbertRTreeAuxiliaryInformation` and `XTreeAuxiliaryInformation`.